docs/writing-rules/data-flow/taint-mode.md

```diff
@@ -39,63 +43,208 @@ You can find more examples of taint rules in the [Semrep Registry](https://semgr
+To specify that some argument of a function must be considered a taint source, simply write a pattern that matches that argument:
 
-In some codebases this assumption may produce too many false positives—in this example, `some_safe_function` may not be returning tainted data after all. If that is the case, you could first consider enumerating such functions as sanitizers (which in a sense they are):
+```javascript
+pattern-sources:
+  - patterns:
+    - pattern-inside: function ($REQ, ...) {...}
+    - focus-metavariable: $REQ
+```
 
-<iframe src="https://semgrep.dev/embed/editor?snippet=jwey" border="0" frameBorder="0" width="100%" height="432"></iframe>
+### Sources by side-effect
 
-If this is too cumbersome, then you can easily "turn it around" and instead assume that every function call is a sanitizer by default:
+Consider the following hypothetical Python code where we call a `make_tainted` function that makes its argument tainted by side-effect:
 
-<iframe src="https://semgrep.dev/embed/editor?snippet=y1jZ" border="0" frameBorder="0" width="100%" height="432"></iframe>
+```python
+def make_tainted(a_set):
+  a_set.add(tainted)
 
-For convenience, Semgrep has a special kind of _not-conflicting_ sanitizer for this purpose, declared with `not_conflicting: true`. A pattern like `$F(...)` matches every function call. If it were acting as a regular sanitizer, it would also apply to any source or sink that had the same function-call shape. In this example, this would sanitize all calls to `sink` and [Semgrep would produce no findings at all](https://semgrep.dev/s/rYrj).
+make_tainted(my_set)
+sink(my_set)
+```
 
-If you use this approach, you instead have to enumerate your taint propagators—with `pattern-not`—if there are any:
+If we wrote the source specification below, the pattern formula would only match the occurrence of `x` that is the actual parameter of `make_tainted`, but it would not affect the occurrence of `x` that is passed to `sink`. They are the same variable, but they are different occurrences, and only the fist one is matched by the source specification. At the same time, Semgrep does not know that `make_tainted` is actually updating the variable `x` by side-effect. Thus, a taint rule using such specification would not produce any finding.
 
-<iframe src="https://semgrep.dev/embed/editor?snippet=9rJ2" border="0" frameBorder="0" width="100%" height="432"></iframe>
+```yaml
+pattern-sources:
+- patterns:
+  - pattern: make_tainted($X)
+  - focus-metavariable: $X
+```
 
-Semgrep also considers that indexing an array with a tainted index leads to tainted data. Again, it is easy to disable this via sanitizers:
+Until recently, the recommended workaround has been to write a source specification like the one below. This tells Semgrep that _every_ occurrence of `$X` after `make_tainted($X)` must be considered a source.
```

Comment:
Note (of course this is TW work, just so you comprehend the reasoning): We also should not use words such as "until recently" in docs - the document may not be updated for the next half a year and it will stay there...